doc/smart-switch/high-availability/smart-switch-ha-hld.md

```diff
@@ -0,0 +1,2176 @@
+
+And to provide HA for DPU-to-DPU communication, we can also leverage our ECMP network setup.
+
+##### 4.2.3.1. HA Control Plane Channel
+
+###### 4.2.3.1.1. HA control plane control channel
+
+The control plane control channel is used for each DPU to talk to each other for things like state machine management. Implementation-wise, we will need a new service for HA communication and management, likely in its own container or SWSS.
+
+Having our own control channel is important. If we consider the switches as data plane and our upstream services as control plane, then data plane should be able to operate on its own, when control plane is disconnected. For example, when control plane is down and network is having trouble, we should still be able to react and adjust our services. Coupling data plane with control plane will introduce more failure modes also making the data plane unable to react properly.
+
+The DPU state machine management will run on NPU. Since each NPU manages multiple DPUs, all NPUs need to connect to all other NPUs, which forms a full mesh.
+
+The data path of this channel will look like below:
+<p align="center"><img alt="HA control plane control channel data path" src="./images/ha-control-plane-control-channel-data-path.svg"></p>
+
+Since DPU events are still coming from the DPU side via SAI/syncd, we will need to pass these events to the NPU side via ZMQ, which go through the PCIe bus.
+
+###### 4.2.3.1.2. HA control plane data channel
+
+The data channel is used for transferring large trunks of data, and being implement on DPU (syncd), because we don’t want these data being passed around, especially on PCIe bus. 
+
+For example, the control channel can send messages which eventually causes bulk sync to be initiated on DPU. However, the bulk sync packets will not go through control channel, but data channel.
+
+This helps us avoid head-of-queue blocking for control messages. Also, we can leave the data channel running on DPU while control channel running on NPU, which allows us to run state machine management independently.
+
+###### 4.2.3.1.3. HA control plane channel data path HA
+
+Although it might be counter-intuitive, we could consider that we can always get a working connection within a bounded time. This is achieved by leveraging our ECMP setup on T1s.
+
+The approach is simply brute force: as long as we have tried enough number of connections with different source port, we can get the channel recovered.
+
+This approach works because with our current setup, it is extremely unlikely that 2 switches cannot communicate with each other. Say, if the possibility of 1 link failing is P, then the possibility of 2 T2s failing talk to each other will be (2p – p^2)^20 to (2p – p^2)^40.
+
+To reduce the time of finding a good path and meet our SLA requirement, we could also try spray the network with a specific number and choose the first one that works.
```

Comment:
By "spray the network," do you mean the CPU uses some algorithm to systematically try a range of ports (i.e. change the next hop egress port) one at a time, until a working port is found? The term "spray" can also refer to one possible mode of a switching ASIC's ECMP logic, where the CPU assigns a port group and the ASIC randomly chooses an egress port for each packet.

Reply:
> The term "spray" can also refer to one possible mode of a switching ASIC's ECMP logic, where the CPU assigns a port group and the ASIC randomly chooses an egress port for each packet.

ouch! I mean the former one. do you have any recommended term to use here?

(btw, I have updated this section as part of the HA control plane data channel redesign, it will be section 5.2.2.2. I hope things will be more clear.)

Reply:
I don't know of a better unambiguous term, but I think using "rotation through available ports using a suitable algorithm in the controller" is clear enough.

Reply:
I have updated it to this: Additionally, to reduce the time of finding a good path and meet our SLA requirement, we could also try to use multiple different source port to establish the new connection in parallel, and choose whichever one that works first.